internal/shared/controllers/pull_secret_controller.go

```diff
@@ -0,0 +1,228 @@
+}
+
+func (r *PullSecretReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx).WithName("pull-secret-reconciler")
+
+	logger.Info("processing event", "name", req.NamespacedName)
+	defer logger.Info("processed event", "name", req.NamespacedName)
+
+	secrets := []*corev1.Secret{}
+	secret := &corev1.Secret{}
+
+	if r.SecretKey != nil { //nolint:nestif
+		// Add the configured pull secret to the list of secrets
+		if err := r.Get(ctx, *r.SecretKey, secret); err != nil {
+			if apierrors.IsNotFound(err) {
+				logger.Info("secret not found", "name", r.SecretKey)
+			} else {
+				logger.Error(err, "failed to get Secret", "name", r.SecretKey)
+				return ctrl.Result{}, err
+			}
+		} else {
+			logger.Info("global pull secret", "name", *r.SecretKey)
+			secrets = append(secrets, secret)
+		}
+	}
+
+	// Grab all the pull secrets from the serviceaccount and add them to the list of secrets
+	sa := &corev1.ServiceAccount{}
+	logger.Info("serviceaccount", "name", r.ServiceAccountKey)
+	if err := r.Get(ctx, r.ServiceAccountKey, sa); err != nil { //nolint:nestif
+		if apierrors.IsNotFound(err) {
+			logger.Info("serviceaccount not found", "serviceaccount", r.ServiceAccountKey)
+		} else {
+			logger.Error(err, "failed to get serviceaccount", "serviceaccount", r.ServiceAccountKey)
+			return ctrl.Result{}, err
+		}
+	} else {
+		nn := types.NamespacedName{Namespace: r.ServiceAccountKey.Namespace}
+		pullSecrets := []types.NamespacedName{}
+		for _, ips := range sa.ImagePullSecrets {
+			nn.Name = ips.Name
+			// This is to update the list of secrets that we are filtering on
+			// Add all secrets regardless if they exist or not
+			pullSecrets = append(pullSecrets, nn)
+			secret := &corev1.Secret{}
+			err = r.Get(ctx, nn, secret)
+			if err != nil {
+				if apierrors.IsNotFound(err) {
+					logger.Info("serviceaccount pull secret not found", "secret", nn)
+				} else {
+					logger.Error(err, "failed to get serviceaccount pull secret", "secret", nn)
+					return ctrl.Result{}, err
+				}
+			} else {
+				secrets = append(secrets, secret)
+			}
+		}
+		// update list of pull secrets from service account
+		logger.Info("updating list of pull secrets", "list", pullSecrets)
+		r.ServiceAccountPullSecrets = pullSecrets
+	}
+
+	if len(secrets) == 0 {
+		return ctrl.Result{}, r.deleteSecretFile(logger)
+	}
+	return ctrl.Result{}, r.writeSecretToFile(logger, secrets)
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *PullSecretReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	_, err := ctrl.NewControllerManagedBy(mgr).
+		For(&corev1.Secret{}).
+		Named("pull-secret-controller").
+		WithEventFilter(newSecretPredicate(r)).
+		Build(r)
+	if err != nil {
+		return err
+	}
+
+	_, err = ctrl.NewControllerManagedBy(mgr).
+		For(&corev1.ServiceAccount{}).
+		Named("service-account-controller").
+		WithEventFilter(newNamespacedPredicate(r.ServiceAccountKey)).
+		Build(r)
```

Comment:
I don't think I'd setup two separate controllers. IIRC, there's a way to have a single controller with multiple watches. You may need to drop down to the lower-level controller package though (can't remember if `For` is required).

Reply:
TBH, having two controllers actually seems the cleanest, in that we get different invocations that recognize ServiceAccount vs. Secret, and it's clear what triggered the reconcile (i.e. the logger clearly indicates what resource is the trigger). With a single controller, we have to somehow map the ServiceAccount to a set of Secrets (which we may not yet know about yet), which feels kinda kludgy. Otherwise, the Secret controller is getting a ServiceAccount for the request, rather than a Secret.